Drop PCP configuration from Overview, add pcp config dialog to Metrics page

martinpitt opened the pull request:
nDrop installation/enablement of pmlogger/pmcd. This functionality moved
to the Performance Metrics page, where it is better motivated and
contextual.

 - [x] Add UI for re-disabling pmlogger to metrics page: #14803
 - [x] test that pmlogger.service gets enabled/disabled (not just stopped/started)
 - [x] `TestHistoryMetrics.testBasic` flake variant [one (4 times)](https://logs.cockpit-project.org/logs/pull-15810-20210517-134329-7d702193-centos-8-stream/log.html#277), [two](https://logs.cockpit-project.org/logs/pull-15810-20210517-134329-7d702193-ubuntu-2004/log.html#268), [three](https://logs.cockpit-project.org/logs/pull-15810-20210517-134329-7d702193-fedora-34/log.html#277-2)
 - [x] Dialog design updates from Garrett

## Metrics: PCP configuration dialog

The Metrics page header has a new "Metrics settings" button:

![metrics-menu](https://user-images.githubusercontent.com/10246/119635105-a5306380-be13-11eb-9262-b544ee29391c.png)

This allows you to enable or disable the [Performance Co-Pilot](https://pcp.io/) (PCP) metrics collector:

![metrics-settings](https://user-images.githubusercontent.com/10246/119635121-a8c3ea80-be13-11eb-9422-e9e8e138ac83.png)



Related options will be added in the near future.

## Other screenshots for PR review
Off setting. I can't say I like the gray "Collect metrics" text, it makes the switch look disabled. But I suppose we should change that globally if we care

![image](https://user-images.githubusercontent.com/200109/118648554-d04aff80-b7e2-11eb-9e5f-88638a6cc14b.png)

Mobile view:

![image](https://user-images.githubusercontent.com/200109/118648724-025c6180-b7e3-11eb-8175-e3d3192947a1.png)

The same button/dialog (except for primary instead of secondary) is also available from the empty state, for a better mental/logical connection. It subsumes the previous "Enable PCP metrics collector" button from that empty state, and the on/off switch on the Overview configuration card:

![image](https://user-images.githubusercontent.com/200109/118499574-5b66bf80-b727-11eb-8742-3b71535fba71.png)

This provides a space where to add more PCP settings. The one that will follow immediately is pmproxy, for Grafana integration (PR #15812), which could look somewhat like this:

![image](https://user-images.githubusercontent.com/200109/118652398-d80ca300-b7e6-11eb-8ced-001453965188.png)

But we could also configure pmcd, pmie, etc. there in the future.

marusak commented:
https://github.com/cockpit-project/cockpit/issues/14803

martinpitt commented:
I pushed a first version of the config dialog. It is functionally correct now, but there is zero layout in the dialog, thus it is very broken:

![image](https://user-images.githubusercontent.com/200109/118480679-dd98b900-b712-11eb-80fd-002cebc379c0.png)

(also needs tests adjustments)

martinpitt commented:
@garrett : I still want to refine the tests, but I'd appreciate a first look at the visuals. I made sceenshots of the new things, they ought to suffice for the first round? I have a gut feeling that we should explain things a little better perhaps?

garrett commented:
Dialog should use `sm` size. The offsite link needs more spacing and the text should not break (it should not wrap, unless absolutely necessary; the link should be considered its own widget). Labels should be to the right of switches; we don't need to do a form layout here with labels on the left (or above). I don't think the .service names are necessary in this dialog. We might want to include more information about the port and/or how to contact the service, however.

![metrics-settings](https://user-images.githubusercontent.com/10246/118508313-5d348100-b72f-11eb-9139-68d8bfffd9bc.png)

martinpitt commented:
Thanks @garrett! Good points, I'll rework this. My [initial attempt](https://github.com/cockpit-project/cockpit/pull/15810#issuecomment-842243498) was indeed without a form, but that looked rather sad -- I mostly want this for the alignment, not for the "form"-i-ness. But I'll revisit this, in other places we e.g. use a DataList, not a form.

> I don't think the .service names are necessary in this dialog.

I thought it would be nice to give a hint that it's a systemd unit. Perhaps "Collect metrics (`pmlogger.service`)"?

garrett commented:
> I thought it would be nice to give a hint that it's a systemd unit. Perhaps "Collect metrics (pmlogger.service)"?

I actually had this as a mockup too. Here's what it looked like:

![metrics-2](https://user-images.githubusercontent.com/10246/118510381-3414f000-b731-11eb-9cf0-3af126300d41.png)

And had a stacked version as well, trying to get rid of the ()

![metrics-3](https://user-images.githubusercontent.com/10246/118510533-5575dc00-b731-11eb-9083-5af48ee05139.png)

Ultimately, I thought it didn't provide much value and dropped the service names altogether.

(At one point, I did even color them blue and was thinking of having them serve as links... but then going to another page within Cockpit via a link _on a modal_ is weird.)

martinpitt commented:
I fixed the testBasic regression. In exchange I broke `TestHistoryMetrics.testNoDataEnable` (I think), but that's a puzzle for tomorrow.

martinpitt commented:
I updated the dialog design as per @garrett's proposals above, I moved from a `<Form>` to `<Flex>`. I'm not 100% sure if that is the right choice, but I didn't find anything more appropriate in PF. (Aside from using a classic plain `<table>`). I updated the screenshots in the description.

I also updated/fixed the tests to what I had in mind, so this is ready for another review round now.

martinpitt commented:
/packit test

KKoukiou commented on pkg/metrics/metrics.jsx:
```diff
@@ -1213,14 +1213,17 @@ class MetricsHistory extends React.Component {
-              </Breadcrumb>
-          }>
+    <Page>
+        <PageSection variant={PageSectionVariants.light}>
```
Breadcrumb should be passed like this https://github.com/cockpit-project/cockpit/blob/master/pkg/storaged/details.jsx#L129
Or you can see also here https://www.patternfly.org/v4/components/page/react-demos but there is so much other junk.

martinpitt replied:
That's exactly how it was before, [this commit](https://github.com/cockpit-project/cockpit/pull/15810/commits/c56aa0f42d945256f4b4aeac2ce7dd7b750af258) moves away from it. See the [discussion in #14803](https://github.com/cockpit-project/cockpit/issues/14803#issuecomment-839703508) for details. Reason 1: This makes the header bar exactly as high as the Overview, so it's more consistent; Reason 2: The Page.breadcrump attribute has no way to place an additional action into the header bar.

KKoukiou replied:
I see, makes sense as a change then.

KKoukiou commented on pkg/metrics/metrics.jsx:
```diff
@@ -867,6 +872,91 @@ class MetricsHour extends React.Component {
+                <Text component={TextVariants.p} className="pcp-settings-modal-text">
+                    { _("Performance Co-Pilot collects and analyzes performance metrics from your system.") }
+
+                    <a href="https://pcp.io/" target="_blank" rel="noopener noreferrer">
```
You can also use PF4's `<Button component='a' variant='link icon=/...>` I don't think it will change the style though.

martinpitt replied:
Thanks! I applied your fixup.

KKoukiou reviewed (commented):
I want to double check the the custom CSS introduced can't be avoided. Will check that locally shortly, pls keep the PR open.

KKoukiou commented on pkg/metrics/metrics.jsx:
```diff
@@ -867,6 +872,91 @@ class MetricsHour extends React.Component {
+                    </a>
+                </Text>
+
+                <Flex direction={{ default: 'column' }}>
```
Using the flex here results in misalignments.
![Screen Shot 2021-05-19 at 12 44 51](https://user-images.githubusercontent.com/14921356/118638216-e18e0f00-b7d6-11eb-8115-8531918fc89a.png)
 

I will try to find a better apporach.

KKoukiou replied:
In addition the Switch should have a label or an aria label to be accessible. Clicking on the label should also enable/disable the switch.

martinpitt replied:
I could also use the "builtin" label of a switch, which will most probably fix the alignment, and certainly fix the "linking" of the two. The main downside is that the label becomes light grey when the button is off, and thus the widget looks disabled when it's not. But that's a general problem of the switches, so let's rather fix that with CSS if we care. Then I just need to have a way to add the (unitname) bit.

I'll make a version with that, and re-post.

KKoukiou replied:
Right, that should be fixed globally is patternfly overrides file if we decide it's wrong.

martinpitt replied:
Thanks for your fixup! applied.

martinpitt commented:
Thanks @KKoukiou for your help! PR and screenshots updated.

KKoukiou commented on pkg/metrics/metrics.jsx:
```diff
@@ -1213,14 +1213,17 @@ class MetricsHistory extends React.Component {
-              </Breadcrumb>
-          }>
+    <Page>
+        <PageSection variant={PageSectionVariants.light}>
```
I see two drawbacks with this change still:
* breadcrumb is not longer sticky
* we dont use the default paddings, indeed overview and kdump use the same approach, but all other pages not. 

Definitely not blockers.

KKoukiou replied:
https://paste.centos.org/view/6c0c0993
This change will keep the header sticky.

martinpitt replied:
Nice, thank you! Admittedly I didn't even notice the non-stickyness. Applied.

KKoukiou reviewed (commented):
Now that I rethink about it the undoing of the sticky header, is a small regression. What do you think?

garrett commented:
FWIW, here's a screenshot:

![Screenshot from 2021-05-19 09-50-52](https://user-images.githubusercontent.com/10246/118775999-caf3c080-b887-11eb-9710-be1eee88c9ff.png)

I only see one option; I guess that's intentional for now?

The read more link is super odd. It links to the PCP website? Is that a placeholder? I expected to read more about PCP works in Cockpit, not just go to the PCP website.

martinpitt commented:
@garrett : Yes, it's intentional. See the description, the "pmproxy" stuff is a separate PR. That was just a preview how the dialog would extend to more options in the future.

The link goes to https://pcp.io/ . It's similar to the other external links that we have, like https://access.redhat.com/security/vulnerabilities/L1TF on the hwinfo side. I thought it'd be nice to link to the actual documentation of the technology that provides all these metrics (the learning Linux/extend to other machines aspect). Neither pmlogger nor pmproxy/Grafana are cockpit specific, it just helps you to set these up. (In the same way as you'd format a drive through udisks on the Storage page, or start a libvirt VM on Machines page).

An alternative link target would be https://cockpit-project.org/guide/latest/feature-pcp.html , which would be one level of indirection -- but maybe that's preferable?

garrett commented:
"Read more" states the page will be more about the text it's talking about. The PCP website isn't that. Our page is.

So we could link our site with more information, adjust the text to make it clear it's just linking to PCP, change the link to be the "Performance Co-Pilot (with the external link icon), or drop the link.

martinpitt commented:
> "Read more" states the page will be more about the text it's talking about. 

Interesting, that's exactly what I had in mind -- the sentence before explains what PCP is, but that dialog is way too small to have all the details about it.

> So we could link our site with more information

OK, I'll link to our feature page then, that seems like the best compromise. I really want some doc link in there.

Thanks!

martinpitt commented:
> I'll link to our feature page then, that seems like the best compromise. I really want some doc link in there.

Done. No visual difference, just the link target changed. @garrett, PTAL?

martinpitt commented:
Pushed a rebase to get past the new "protection checks".

garrett commented on pkg/metrics/metrics.jsx:
```diff
@@ -867,6 +872,92 @@ class MetricsHour extends React.Component {
+                       { dialogError && <ModalError dialogError={ _("Failed to configure PCP") } dialogErrorDetail={dialogError} /> }
+
+                       <Button variant='primary' onClick={handleApply} isDisabled={pending}>
+                           { _("Apply") }
```
```suggestion
                           { _("Save") }
```

This should be "Save", not "Apply".

garrett reviewed (changes requested):
Looks good!

The action button should say "Save" though. Aside from that, it's ready.

garrett reviewed (approved):
Cool! I see you've changed it to save. :+1:
